File: leetcode/c++/easy/_1455_check_if_a_word_occurs_as_a_prefix_of_any_word_in_a_sentence.cpp
// Leetcode Problem 1455

#include <gtest/gtest.h>
#include <string>
#include <sstream>
#include <vector>
#include <regex>

using namespace std;

class Solution {
public:
    int isPrefixOfWord(string sentence, string searchWord) {
        int i = -1, wordIndex = 0;

        do {
            ++i, ++wordIndex;
            if (sentence.substr(i, searchWord.size()) == searchWord) {
                return wordIndex;
            }
        } while ((i = sentence.find(' ', i)) != string::npos);

        return -1;
    }

    /**
     * Solution i came up with but is very slow
     */
    int isPrefixOfWord_regexMatching(string sentence, string searchWord) {
        if (sentence.length() < searchWord.length()) {
            return -1;
        }

        int position = 1;

        string word;
        istringstream stream(sentence);

        string escapedWord = regex_replace(searchWord, regex("([\\^\\.\\$\\|\\(\\)\\[\\]\\\\\\+\\*\\?\\{\\}])"), "\\$1");
        string pattern = "^" + escapedWord;

        regex regexPattern(pattern);

        while (stream >> word) {
            if (regex_search(word, regexPattern)) {
                return position;
            }

            position++;
        }

        return -1;
    }
};


class TestClass : public ::testing::Test {
protected:
    Solution solution;
};

TEST_F(TestClass, SimplePossibleCase) {
    string sentence = "i love eating burgers";
    string seachWord = "burg";

    EXPECT_EQ(solution.isPrefixOfWord(sentence, seachWord), 4);
}

TEST_F(TestClass, SimplePossibleCase2) {
    string sentence = "this problem is an easy problem";
    string searchWord = "pro";

    EXPECT_EQ(solution.isPrefixOfWord(sentence, searchWord), 2);
}

TEST_F(TestClass, ImpossibleCase) {
    string sentence = "i am tired";
    string searchWord = "you";

    EXPECT_EQ(solution.isPrefixOfWord(sentence, searchWord), -1);
}

TEST_F(TestClass, SearchWordTooLong) {
    string sentence = " ";
    string searchWord = "you";

    EXPECT_EQ(solution.isPrefixOfWord(sentence, searchWord), -1);
}

TEST_F(TestClass, EmptySentence) {
    string sentence = " ";
    string seachWord = "a";

    EXPECT_EQ(solution.isPrefixOfWord(sentence, seachWord), -1);
}

File: leetcode/rust/problems/src/easy/_1455_check_if_a_word_occurs_as_a_prefix_of_any_word_in_a_sentence.rs
// Leetcode Problem 1455

use regex::Regex;

struct Solution;

#[allow(dead_code)]
impl Solution {
    pub fn is_prefix_of_word(sentence: String, search_word: String) -> i32 {
        if sentence.len() < search_word.len() {
            return -1;
        }

        let words = sentence.split(" ");
        let escaped_word = regex::escape(&search_word);
        let pattern = format!("^{}", escaped_word);
        let regex = Regex::new(&pattern).unwrap();

        for (i, word) in words.enumerate() {
            if regex.is_match(word) {
                return (i + 1) as i32;
            }
        }
        
        -1
    }
}


#[test]
fn test_is_prefix_of_word() {
    let sentence1 = String::from("i love eating burgers");
    let sentence2 = String::from("this problem is an easy problem");
    let sentence3 = String::from("i am tired");

    assert_eq!(Solution::is_prefix_of_word(sentence1, String::from("burg")), 4);
    assert_eq!(Solution::is_prefix_of_word(sentence2, String::from("pro")), 2);
    assert_eq!(Solution::is_prefix_of_word(sentence3, String::from("you")), -1);

    assert_eq!(Solution::is_prefix_of_word(String::from(" "), String::from("you")), -1);
    assert_eq!(Solution::is_prefix_of_word(String::from(" "), String::from("a")), -1);
}